### DISCUSSION TITLE
Support for vector overlay

### DISCUSSION THREAD

**Discussion by Greydesk** [osmandapp/OsmAnd#13566] [Ideas]
Hi all,
I've been using OSMAND since I degoogled my phone and I'm very impressed with all its features. I use it for so many things in driving and hunting. I do have one request: please implement some sort of vector format for overlays. I have access to the crown (public) land data in my province in many vector formats but OSMAND does not support any of these. I have had to convert to a geotiff, then mbtiles and such before I can use it as an overlay and then it is blurry at any usable zoom. 
This seems like it would be a simple addition since you already support some vector files, just not in the overlay area. GeoJson, Shapefile, etc. would all work, as long as there is one vector format supported.
Thanks!

**Comment by mariush444** [comment]
just workaround but probably you can create obf file.
it is not the best solution but let's say "visible"

**Reply by Greydesk** [reply]
The attempts I have made to create an obf files had failed. Each of them results in the map not updating if you scroll around. It appears that the map rendering locks up. The only solution that has worked is an sqlitedb file. I would love any workable solution.
At this moment, I am using ogr2ogr to attempt to convert the shapefile into an mbtiles database. It has been running since Monday and hasn't made it to 40% yet. Seems to be stuck.
This is the command I am using:
ogr2ogr -f MBTILES -dsco MINZOOM=10 -dsco MAXZOOM=22 -progress /media/statik/Statik/CrownLand/MBTiles/test.mbtiles /media/statik/Statik/CrownLand/geo_export_6f8be9aa-e15c-4de9-92e0-6489e8a4826d.shp 
0...10...20...30...

Thanks for the reply, tho.

**Reply by mariush444** [reply]
generally I created tiles from geotiif, geopdf and ozi.
Shapes I created just for test probably via Qgis as I remember.

But obf I created some time ago with OsmAndMapCreator for some roads and POI in off-roads needs and it rather worked OK.

**Reply by Greydesk** [reply]
I have not been able to get a working obf from OsmAndMapCreator. I would love another set of eyes on the problem.
I can download the overlay data from here: https://data.novascotia.ca/Lands-Forests-and-Wildlife/Crown-Land/3nka-59nz#content
There are many options for file type but I have not been able to get any of them to work. It is definitely a learning curve.:)

**Reply by mariush444** [reply]
I made mistake. So ....
To load geojson or shp you need JOSM. With this tool you can convert (save as) your data to osm.
Then osm can be converted to obf with OsmAndMapCreator

!!! remember - in JOSM, never send uploaded data (layer) to server, it will damage openstreetdata !!!

well... but it is not so easy. The most difficult step is in JOSM.
When you loaded your data to JOSM you should set (let's say) the tag "what it is"

e.g. my curves, I have to set as service road to achieve what I need in osmand.

The easiest method is to use menu "Presets". You should download some of them at the beginning.
Then you have try which one will be the best for you.
I tried to use some boundaries and some land use like cementary but without result. It was not visible in osmand.
You need a lot of test so prepare a small part of your data as sample then try step by step.

At the end you can set this tag for all your data and convert whole set to obf.

**Reply by Greydesk** [reply]
I just tried this using the preset Recreational Ground. Nothing shows up in OSMAND for the file I created. Not in map source or overlay. I guess I have to keep going through the presets until I find one that shows up?

**Reply by mariush444** [reply]
Yes, I think so. I learn on my mistakes too. You can probably create new discussion here or JOSM. Maybe someone with more experience will help.

**Reply by QuestionsForThose** [reply]
You might want to look at https://www.youtube.com/watch?v=qVuQenIzwik&t=34s
Custom Maps app

**Reply by mariush444** [reply]
good to know but the question was for vector and not raster maps.
definitely there is lack of such features in osmand but
discussion about raster was closed https://github.com/osmandapp/OsmAnd/issues/9944 by osmand team

I prefer  Oruxmap for raster maps (geopdf/tif, ozi)
but it is possble to convert them to osmand 
https://github.com/mariush444/raster2osmand

**Comment by mariush444** [comment]
This
![Screenshot_20220120-145215](https://user-images.githubusercontent.com/66887280/150353114-7f45eea2-64b8-4fbe-b195-05fe9e93031e.png)


I get for settings like that
![image](https://user-images.githubusercontent.com/66887280/150353050-ca43855a-0851-4f23-8407-04d30610f5e9.png)

example 
[Sample-admin.obf.zip](https://github.com/osmandapp/OsmAnd/files/7905748/Sample-admin.obf.zip)

**Reply by Greydesk** [reply]
Thanks, I'll try that.

**Reply by Greydesk** [reply]
Where in the directory structure are you placing the obf file?

**Reply by mariush444** [reply]
the main, the same what you choose in main osmand settings
where all the other obf's file can be found (probably you have Brunshwik)

**Reply by thompehw** [reply]
From reading other discussions/issues here, placing a .obf directly in OsmAnd's folder no longer works if you're on Android 11+. 

Android now prevents OsmAnd from seeing files it did not create itself. If this is your case, try putting the .obf somewhere else (e.g. Downloads) then use a file manager app (e.g. Ghost Commander) to open it and select OsmAnd as the app in which to open it.

From what others have said, this causes the .obf to be copied/imported by OsmAnd so that it can then see it.

**Reply by mariush444** [reply]
https://github.com/osmandapp/OsmAnd/discussions/13539
read at the end, pls

**Reply by Greydesk** [reply]
I am on android 10. I have created an sqlitedb file before that I placed in the tiles folder and it is recognized. I have not had the obf file recognized in main or tiles.
I moved the obf to another location and then did the open-with thing and it copied into the files directory but is still not available as an overlay or a map source.
I'm thinking it has to do with the file itself. I have been playing with QGIS and it reports some errors with the data but I haven't been able to discover how to fix them. The two I found that seem most serious involve a single vector to a point and back to the same point but I haven't been able to delete that single point yet.

**Reply by mariush444** [reply]
I'm not sure I understand your problems.
When you see tiles directory then you can see main directory that contained tiles dir.
Just put there my example and go to New Scotia. That's all. 
You don't need switch on/of over/under lay or map source.

**Reply by thompehw** [reply]
I tested the example file of @mariush444 and it worked as shown (I'm also on Android 10). As mentioned, an .obf is not selected with map_source/overlay/underlay - that's for tiles. The .obf will show by default, but you can deactivate/activate from Download_maps>Local.

I tested in OsmAnd and TouringView map styles, and those renderers display the example features from zoom level 9 and closer.

It might be helpful to state some things that may be assumed as obvious! After importing the data into JOSM from shapefile or whatever, the idea is to add OSM tags to all the objects to make them look like something that OsmAnd usually displays. Then save as .osm and use OsmAndMapCreator to make .obf.

Look at the .obf file size. If it's very small, then OsmAndMapCreator didn't like your tagging.

Finally, supplied with OsmAndMapCreator is the command  line utility inspectot.bat/.sh (windows/linux) which can be used to see what's in the .obf. I have used:
`inspector.bat -vmap -vmapobjects -osm myfile.obf`
to see the objects in a way familiar from OSM.

I think the original request for OsmAnd to support at least one standard vector format is still interesting, but with unknown tags I guess objects would have to be rendered with a uniform default style (e.g.  bold black).

**Reply by Greydesk** [reply]
Thank you to everyone who replied. This is now solved. As thompehw explained, the problem was caused by me not understanding how to add more vector maps to OSMAND. While I am still playing with the options for which preset will give me the look I am going for, the procedure about works amazingly. I can't believe I've been struggling at this for almost a year but the answer was so simple.
Here's some screenshots of where I was and what it looks like now. https://imgur.com/a/E7NLgMr

**Reply by thompehw** [reply]
That's great!

An even easier approach that can be remarkably effective for linear features is just to use JOSM to export to GPX. Import the shapefile/whatever, select all and delete the tags (this is just to make the file smaller). Then export to GPX - no need to add OSM tagging.

OsmAnd is now very sophisticated at handling GPX. You can easily change the appearance of each GPX in your collection, turn them on or off, edit them, or even smooth  them using "GPS filter".

This approach also makes it clear to see that the data is not part of OSM, and it's easy to display features of a type that OsmAnd would not normally render. 

If OsmAnd were to consider adding support for standard vector data for people who don't have access to JOSM, I would suggest looking at simply converting to GPX for these reasons.

**Reply by mariush444** [reply]
GPX - yes and no, in my opinion.
In this case the gpx file is about 190MB and osmand crashes with the file on my device.
Can you verify on yours, pls?

But there is no problem with big obf when it's red partially by osmand. GPX is red at once as I understand.

So for small gpx files - yes, for big ones and for area - no.
Just my observations.

**Reply by thompehw** [reply]
Thanks, @mariush444! I was speaking from previous experience on smaller datasets so had not appreciated the crucial difference that .gpx is fully loaded but .obf is more intelligent.

I had experimented with vector data from the local authorities of English counties that give the legally designated Public Rights of Way. These are much smaller vector files. Even then, to save space on my older device I used the GPSBabel "crosstrack" simplification method with an error parameter of 5-10m. I tried the same approach with the Nova Scotia data, and the parameter had to be 100m before my newer device could (slowly) cope with the GPX. 

So I agree with you!
